### ISSUE TITLE
New LeftCurlyOption: `nl_or_singleline` to allow single-line blocks in otherwise Allman-style code

### ISSUE THREAD

**Issue by willhains** [checkstyle/checkstyle#12226]
I have read check documentation: https://checkstyle.org/config_blocks.html#LeftCurly
I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

**How it works Now:**

```bash
$ javac Example.java 

$ cat MyConfig.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
	"-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
	"https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
	<property name="severity" value="error"/>
	<property name="fileExtensions" value="java"/>

	<module name="TreeWalker">
		<module name="LeftCurly">
			<property name="option" value="nl"/>
		</module>
		<module name="RightCurly">
			<property name="option" value="alone_or_singleline"/>
		</module>
	</module>
</module>

$ cat Example.java 
// Empty brace pair on same line: OK
class SingleLineOK { }

// Empty brace pair on next line: OK
class SingleNextLineOK
{ }

// Empty brace pair on multiple lines: OK
class MultiLineOK
{
}

// Empty brace pair with trailing left curly: FAIL
class MultiLineFAIL {
}

class MethodExamples
{
	// Empty brace pair on same line: OK
	void singleLineOK() { }

	// Empty brace pair on next line: OK
	void singleNextLineOK()
	{ }

	// Empty brace pair on multiple lines: OK
	void multiLineOK()
	{
	}

	// Empty brace pair with trailing left curly: FAIL
	void multiLineFAIL() {
	}

	boolean x;

	// Single statement on same line: OK
	void singleLineStatementOK() { x = true; }

	// Single statement on next line: OK
	void singleNextLineStatementOK()
	{ x = true; }

	// Multiple statements starting on same line: FAIL
	boolean multiStatementSplitLinesFAIL() { x = true;
		return x;
	}

	// Multiple statements with trailing left curly: FAIL
	boolean multiStatementTrailingLeftCurlyFAIL() {
		x = true;
		return x;
	}
}

$ RUN_LOCALE="-Duser.language=en -Duser.country=US"

$ java $RUN_LOCALE -jar checkstyle-10.3.3-all.jar -c MyConfig.xml Example.java 
Starting audit...
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:2:20: '{' at column 20 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:14:19: '{' at column 19 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:20:29: '{' at column 22 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:32:28: '{' at column 21 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:38:38: '{' at column 31 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:45:46: '{' at column 39 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:50:53: '{' at column 46 should be on a new line. [LeftCurly]
Audit done.
Checkstyle ends with 7 errors.
```

**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
A new LeftCurlyOption similar to `nl`, but allowing end-of-line placement if and only if the right curly brace is on the same line.
Combined with RightCurlyOption `alone_or_singleline`, it should accept all the blocks marked "OK", and reject all the blocks marked "FAIL", in the above example Java source code.

Put another way, it should pass the check if the open and close brace pairs are either in the same row, or in the same column; and fail otherwise.

Suggested LeftCurlyOption name: `nl_or_singleline`

Suggested cli output:
```bash
$ java $RUN_LOCALE -jar checkstyle-10.3.3-all.jar -c MyConfig.xml Example.java 
Starting audit...
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:14:19: '{' at column 19 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:32:28: '{' at column 21 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:45:46: '{' at column 39 should be on a new line. [LeftCurly]
[ERROR] /Users/willhains/Java/CheckStyle/Example.java:50:53: '{' at column 46 should be on a new line. [LeftCurly]
Audit done.
Checkstyle ends with 7 errors.
```

Note: the errors for lines 2, 20, and 38 are gone.
They correspond to the following code snippets, which should be allowed with the proposed LeftCurlyOption `nl_or_singleline`.

```java
// Empty brace pair on same line: OK
class SingleLineOK { }
```

```java
// Empty brace pair on same line: OK
void singleLineOK() { }
```

```java
// Single statement on same line: OK
void singleLineStatementOK() { x = true; }
```

**Comment by willhains** [comment]
Another example that should be allowed with `nl_or_singleline` (if OneStatementPerLine is not used):
~~~java
// Multi-block on same line: OK
try { return array[0]; } catch (Exception e) { e.printStackTrace(); }
if (x) { return 1; } else { return 0; }
~~~

**Comment by MazenAmria** [comment]
Is there anything line `eol_or_singleline`? or is there any alternative way to do it?

**Comment by willhains** [comment]
@MazenAmria I have tried several options and combinations of options, but I haven't found any alternative way to do it. If there is one that behaves as described above, I would love to know.

**Comment by MazenAmria** [comment]
@willhains me too, I've tried different combinations and nothing seemed to work.
